### ISSUE TITLE
Markdown `[..]` as alternative to @link - needs decision

### ISSUE THREAD

**Issue by Fr0sT-Brutal** [pasdoc/pasdoc#151]
Markdown block `[..]` could serve as analog to `@link` and could be added easily. However, this change could break some entries that are not meant to be links because these brackets could appear for another purposes, like
```
type
TR = record
  // [optional]
  a: byte;
  b: byte;
end;
```

However, only standalone blocks will be affected, that is, array item declarations `arr[idx]` won't.
Using another block signatures would break sync with `[..](..)` block defining an URL and bring confusion.
Any ideas?

My opinion: Do it. Affected range is quite low and PasDoc will emit warnings on non-found links.

**Comment by michaliskambi** [comment]
I agree. That is,

- It is important to provide a really easy (to use and to remember) Markdown syntax for possibly the most-often used PasDoc tag, `@link`. Using `[MyItem]` as alternative to `@link(MyItem)` looks simple and consistent.

- As it will be only active when [--markdown](https://pasdoc.github.io/MarkdownOption) is used, and we warn about invalid links, the eventual clashes with someone that tries to use `[` without wanting a link are not a big issue. I agree with your arguments fully.

Additions:

- We should support links with explicit "display text" too. https://pasdoc.github.io/LinkTag . To be consistent with Markdown and Doxygen, I propose we allow `[my link text](#MyItem)` as alternative to `@link(MyItem my link text)`. So, when `[xxx]` is followed immediately by `(yyy)`, it should be interpreted differently than just `[xxx]`. The `#` in this case allows to precisely differentiate between linking to some URL vs linking to Pascal item.

- I looked at Doxygen Markdown support: https://www.doxygen.nl/manual/markdown.html . They support `[The link text](#MyClass)`, so we would be consistent with it.

- We should allow escaping `[` with backslash, for both URL and Pascal links. That is,

```
[--markdown](https://pasdoc.github.io/MarkdownOption) - this is converted a link in Markdown 

\[--markdown](https://pasdoc.github.io/MarkdownOption) - this is NOT converted to a link in Markdown (you can test in GitHub)

[MyItem] - this should be converted to @link by PasDoc

\[MyItem] - this should NOT be converted to @link by PasDoc

[my link text](#MyItem) - this should be converted to @link by PasDoc

\[my link text](#MyItem) - this should NOT be converted to @link by PasDoc
```

**Comment by Fr0sT-Brutal** [comment]
I implemented it but #152 must be applied first, GH doesn't allow to create another PR and just puts commits into first one. Or should I just push commits to dev branch without creating PR's?

**Comment by Fr0sT-Brutal** [comment]
Oh, I wasn't aware I can create another branch here!
Adding `[]`-s is upcoming

**Comment by michaliskambi** [comment]
Following https://github.com/pasdoc/pasdoc/pull/154/ :

- I updated docs to describe what we plan: https://pasdoc.github.io/MarkdownOption , https://github.com/pasdoc/pasdoc/wiki/MarkdownOption .

- I updated `tests/testcases/ok_markdown_links.pas` with some test that don't yet work. I know you're working on this, so it's all good, just making a note.

    `Link without explicit title [ok_markdown_links].` - should be a link, not yet.
    
    These are correctly not links, but we should strip the backslash (otherwise user has no easy way to write something visible as just `[` in output):

    `\[Test Procedure](#MyProc)`
    
    ```
    Not links to \[Self](#MyProc), and a \[Second one](#ok_markdown_links.MyProc).
    Not a link without explicit title \[ok_markdown_links].
    ```

    Compare with what happens if you type `\*my file*` in GitHub, even in comments here. The resulting output is `*my file*`, i.e. asterisks do not make italic, but the backslash is stripped.

**Comment by michaliskambi** [comment]
> These are correctly not links, but we should strip the backslash (otherwise user has no easy way to write something visible as just [ in output):

Ah, and you already recorded this in issue https://github.com/pasdoc/pasdoc/issues/153 , cool.

**Comment by Fr0sT-Brutal** [comment]
> should be a link, not yet.

Yes, adding this is piece a cake (just one more block tag) but will require changes in some other markdown tests for Urls. Eating escapes is harder though. Now the tag scanner just skips what it considers non-tag and it has no abilities to modify input. So it seems escape stripping should happen somewhere upper, in the generic input scanner.
